src/ol/interaction/Snap.js

```diff
@@ -31,40 +31,226 @@ import {getUid} from '../util.js';
 import PointerInteraction from './Pointer.js';
 
 /**
- * @typedef {Object} Result
- * @property {import("../coordinate.js").Coordinate|null} vertex Vertex.
- * @property {import("../pixel.js").Pixel|null} vertexPixel VertexPixel.
- * @property {import("../Feature.js").default|null} feature Feature.
- * @property {Array<import("../coordinate.js").Coordinate>|null} segment Segment, or `null` if snapped to a vertex.
+ * @typedef {Array<import("../coordinate.js").Coordinate>} Segment
+ * An array of two coordinates representing a line segment, or an array of one
+ * coordinate representing a point.
  */
 
 /**
  * @typedef {Object} SegmentData
  * @property {import("../Feature.js").default} feature Feature.
- * @property {Array<import("../coordinate.js").Coordinate>} segment Segment.
+ * @property {Segment} segment Segment.
  * @property {boolean} [isIntersection] Is intersection.
  */
 
+/**
+ * @typedef {(geometry: import("../geom/Geometry.js").default, projection?: import("../proj/Projection.js").default) => Array<Segment>} Segmenter
+ * A function taking a {@link module:ol/geom/Geometry~Geometry} as argument and returning an array of {@link Segment}s.
+ */
+
+/**
+ * Each segmenter specified here will override the default segmenter for the
+ * corresponding geometry type.
+ * @typedef {Object} Segmenters
+ * @property {Segmenter} [Point] Point segmenter.
+ * @property {Segmenter} [LineString] LineString segmenter.
```

Comment:
If these properties were nullable, you could disable snapping to some geometry types. The rest of the code seems to already handle geometries without available segmenters.

Reply:
Good suggestion. It's simpler for users than specifying a segmenter that returns an empty segments array. Done, along with making the segmenter generic to get better type hints in custom segmenters.